### DISCUSSION TITLE
Diff for a range selection in a commit list always use begin^..end

### DISCUSSION THREAD

**Discussion by geza-herman** [magit/magit#4643] [Ideas]
Magit's diff range feature is a little bit inconsistent, and this inconsistency has some drawbacks. I understand the current design, but still, maybe a different way of working would be better.

Here's what I mean. If I select a commit range (for example, in the commit list of a branch in the `magit-show-refs` buffer), then depending on the size of the range `begin .. end`, magit does a different thing: if a single commit is selected, then it shows the diff of that commit: `begin^ .. end`. But, if multiple commits are selected, it shows the diff of the selected commits except the first: `begin .. end`. This can be a problem, because if I'd like to check out the diff of all the commits which is available in a list, then I cannot do it (because I'd need to include the commit before the first one). To be able to select the needed range, I need to get a log of the end commit, and then I can select the desired range.

When selecting commits in a commit list for diffing, I think about the selected range as "give me the diff that the selected commits introduced", and not "give me the output of `git diff begin end`". And magit behave like this if I select a single commit, but for multiple commits it doesn't. So the diff of single-commit and a commit-range-of-two does the same thing.

What do you think about always using `begin^..end`?

**Comment by tarsius** [comment]
See #3482, especially the last post.

**Reply by geza-herman** [reply]
It's good to see that someone else mentioned exactly the same things as me.

I understand the problem with graphs. Here's what I think: there are two kind of range diffs that people execute:
 1. There is a commit list (usually a linear thing, each commit is the sole parent of the next one), and some of the commits are selected to get the diff that the selected commits introduced. Here, the visual selection is intuitive (for me, at least).
 2. There are two commit hashes, and we need the diff of these two states. This is what `git diff begin .. end` does. Here, I wouldn't say that selecting the range between the two hashes are intuitive. A more intuitive way would be that the user had to select the begin and end point separately (because the range itself doesn't have any meaning, the commits listed between the two hashes don't have a meaning). But I understand that selecting the begin and endpoint is faster with selecting the range.

(Maybe it would help if magit supported some way to extend the range after a diff is executed?)

Thanks for your support, I think I'll hack together something which fits my needs.
